### PR TITLE
Hooks: Use Conda's metadata files to identify binary dependencies.

### PR DESCRIPTION
**Pull request by bwoodsend**
I explain more fully in the commit message but this PR adds a new `conda_support` submodule to `PyInstaller.utils.hooks` which parses some metadata jsons stored inside a Conda installation. These jsons are like the Conda package equivalent of a `MANIFEST.in` in that it lists all the files belonging to a particular package. This will hopefully replace our disastrous attempts to find DLLs for packages such as numpy which Conda has heavily altered.

This is still very much a draft so no need to review it in full yet. But I'd like everyone's thoughts on some of the concepts. Namely:
- Should this be only a hook utility? Or should we try to integrate it into PyInstaller's analysis by default?
- Currently I've put all the API into its own module which can only be accessed if in a Conda environment. This is mainly so I don't have to write placebo functions and/or wrap large definitions in `if` statements for the case of not being in a Conda environment, where all these methods shouldn't be used. I'm not wildly fond of this - what do other people think?

I've briefly added some docs to go with it - more just to explain what I'm doing at the moment. I realise the rest of the docs doesn't use `autodoc` and what I've added using `autodoc` sticks out as being in a different format. But I'll fix that later...

### CODE REVIEW

File: PyInstaller/utils/hooks/conda.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            # Read the json to check we've found the right one.
+            # XXX: Maybe we should make this case and '-'/'_' insensitive?
+            return dic
+    raise ModuleNotFoundError("Distribution {} is either not installed or was"
+                              " not installed using Conda.".format(name))
+
+
+def walk_dependency_tree(initial, excludes=None):
+    """
+    Recursively walk the dependencies of a distribution.
```
Comment (Legorooj) [inline PyInstaller/utils/hooks/conda.py:92]: This function uses a work stack, it isn't actually recursive. I know this is a nitpick, but I like things being technically correct.

File: PyInstaller/utils/hooks/conda.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+             is the output of ``search_distribution(distribution_name)``.
+    :rtype: dict
+
+    """
+    if excludes is not None:
+        excludes = set(excludes)
+
+    # Rather than use true recursion, mimic it with a to-do queue.
+    distributions_done = {}
+    names_to_do = {initial}
```
Comment (Legorooj) [inline PyInstaller/utils/hooks/conda.py:108]: Have you considered using a `collections.deque` for this? They're designed for situations like this.

**Review by Legorooj — CHANGES_REQUESTED** [review]
This is stellar work. I'd like @bjones1 to take a look as well though.

I'm looking forward to the huge amount of bugs this will fix - at least till conda decides to change how it's metadata files work of course :joy:

File: doc/conf.py [hunk 3]
```diff
@@ -64,7 +64,8 @@
 # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
 # ones.
 extensions = ['sphinx.ext.intersphinx',
-              'pyi_sphinx_roles']
+              'pyi_sphinx_roles',
+              'sphinx.ext.autodoc',]
```
Comment (Legorooj) [inline doc/conf.py:68]:
```suggestion
              'sphinx.ext.autodoc']
```
Appease the linter.

**Review by Legorooj — CHANGES_REQUESTED** [review]

File: PyInstaller/utils/hooks/conda.py (hunk 2 above)
Comment (bwoodsend) [reply PyInstaller/utils/hooks/conda.py:108]: You can't efficiently use `in` with a deque so keeping the to-do list duplicate free will become expensive.

Comment (bwoodsend) [reply PyInstaller/utils/hooks/conda.py:108]: Actually... it doesn't matter if there are duplicates in the to-do list. The `if _name in distributions_done: continue` will prevent any real work being run more than once. Deque it is then...

File: doc/conf.py (hunk 3 above)
Comment (bwoodsend) [reply doc/conf.py:68]: Well I think I'm going to revert this because the rest of the docs doesn't use `autodoc`. Although I wish it did...

Comment (Legorooj) [reply doc/conf.py:68]: @bwoodsend no, don't revert this. I'd like to transition to autodoc in the long run.

**Review by bjones1 — COMMENTED** [review]
I don't use Conda, so I can't really provide any meaningful review.

It would be great if the Conda team could somehow maintain this, but that's another complex question since this isn't a hook and there's no nicely-defined extension mechanism.

**Comment by rokm** [comment]
This is going to go a long way towards proper collection of dynamic libraries in conda environments.

>  Should this be only a hook utility? Or should we try to integrate it into PyInstaller's analysis by default?

I suspect a hook utility will be more practical, because I don't imagine analysis will handle package-name vs. import-name discrepancies, and probably the best place to provide mappings between the two is in hooks.

> Currently I've put all the API into its own module which can only be accessed if in a Conda environment. This is mainly so I don't have to write placebo functions and/or wrap large definitions in if statements for the case of not being in a Conda environment, where all these methods shouldn't be used. I'm not wildly fond of this - what do other people think?

Separate module that's available only in Conda environment sounds reasonable.

But I don't think the currently documented try-catch approach to importing it will look good in our hooks, which are mostly based on if blocks for different platforms. Perhaps we could have a `is_pure_conda` bool to indicate that the module can be imported? (Or repurpose the `is_conda` to indicate a pure conda environment; I don't think using `venv` inside conda virtual environments makes much sense, and mixing conda and pip is likely going to lead to a disaster anyway...)

**Comment by bwoodsend** [comment]
> Perhaps we could have a `is_pure_conda` bool to indicate that the module can be imported? (Or repurpose the `is_conda` to indicate a pure conda environment;

I like the `is_pure_conda` bool idea. I'll switch it to that.

> I don't think using `venv` inside conda virtual environments makes much sense, and mixing conda and pip is likely going to lead to a disaster anyway...)

Ha, back in the dark days when I used Conda by default, this was the only solution I had to a lot of problems. I certainly wouldn't write it off as never going to happen - even if it does seem like a nonsense thing to do.

**Review by htgoebel — CHANGES_REQUESTED** [review]
Thanks for this huge work. I totally agree in not putting too much effort here (so same time or space), since Conda is a commercial product and they should come along with a solution.

Anyway I wonder 
1. is there some ticket for discussion the aims and design of this change or is this all in the pull-request?
2. When checking some packages in https://repo.anaconda.com/pkgs/main/win-64/, I could not spott these directory and filenames you mention in the commit message. Does conda create these when installing the pacakge?
3. If conda changes the filename, how can the extension module be imported at all in Conda?
4. how we could get rid of adding a hook for every package
5. How to get from the package name or file to the distribution name - see proposal below.
6. whether this could be written to be more general and support non-conda installations, too
7. Why do we need to process the requirements of each package?

**Regarding 4 & 5**: Implementing a hook for every package containing binary modules is cumbersome. AFAIU the commit message, extending some search paths does not help since conda renames the files.

There is a module `importlib.metadata` (new, in Python 3.8, backported to 3.4-3.7) as `importlib_metadata`), we could use.

Anyway we could get the distribution name from a module-filename with something like this (pseudo-code):
```
file_distribution_map = dict(
   (str(fp), d.metadata["name"])
   for d in importlib.metadata.distributions()
   for fp in d.files)

dist_names = set(  # distributions to consider
    file_distribution_map[node.filename]
    for node in graph.flatten(start=graph._top_script_node))
```
Yes, this is ugly, but since it is just for conda, I do not care much.

**And** there is already building.toc_conversion.DependencyProcessor, which looks like being a good plae for adding this.

**Regarding 6**: Our code should implement the same interface as `importlib.metadata` for reading Conda meta-data. This might allow removing our own code once Conda provides a proper implementation of `importlib_metadata` - or if Conda provides meta-data following PEP 566.

Example
```
>>> from importlib_metadata import requires, files, version
>>> requires("requests")
['chardet (>=3.0.2,<4)', …]
>>> files("pillow")
[…, PackagePath('Pillow.libs/libXdmcp-e15573e7.so.6.0.0'), …]
>>> version('pillow')
'7.2.0'
```

**Comment by bwoodsend** [comment]
Thanks, for the feedback. To answer your questions:

>     is there some ticket for discussion the aims and design of this change or is this all in the pull-request?

Not generally no. There are quite a collection of issues about specifically about numpy, which is probably the package Conda have modified/butchered the most. I had quite a rant [in 5076](https://github.com/pyinstaller/pyinstaller/issues/5076#issuecomment-683436017) about why Conda support is so impractical with our current approach. I suppose that outlines the aims.

>     When checking some packages in https://repo.anaconda.com/pkgs/main/win-64/, I could not spott these directory and filenames you mention in the commit message. Does conda create these when installing the pacakge?

They're not part of the package. I'm not sure where they come from exactly but I think there must a big online database of them somewhere. Conda pulls them in whenever you install something so it can *solve* your environment. 

>     If conda changes the filename, how can the extension module be imported at all in Conda?

I'm not sure what you're referring to. Extension modules aren't renamed. DLLs are moved and some new DLLs become dependencies - that's all.

>     how we could get rid of adding a hook for every package

Not every Conda package really needs this. Only those that Conda have actively altered. Packages that authors have dumped as-is onto conda-forge to stop Conda users complaining about having to use pip will preserve their usual structure and our regular `collect_dynamic_libs` will function fine. And there appears to be a mass effort into creating conda packages out of libraries that don't do it themselves ([including us](https://anaconda.org/conda-forge/pyinstaller)); these are also relatively unchanged. Only the big libraries get the special Conda treatment. That all being said - if we can make this automatic, I definitely think we should.

>     whether this could be written to be more general and support non-conda installations, too

Do you mean, *could we do something similar using `importlib.metdata` to find DLL dependencies when not using Conda*?

>     Why do we need to process the requirements of each package?

I was hoping that was made clear [here](https://github.com/pyinstaller/pyinstaller/pull/5213/commits/a25a17828d6f34dd7e9fa7318f7eadf9197f2f68#diff-65e6be65f569ed3e9dc8dad3609bd082R67-R71). Conda introduce a lot of intermediate packages and the DLLs are actually dependencies of those.

```python
# Numpy itself has no DLL dependencies.
>>> collect_dynamic_libs("numpy", recurse_dependencies=False)
[]
# In fact it is a empty package.
>>> search_distribution("numpy")["files"]
[]
# But... it depends on `numpy-base`.
>>> search_distribution("numpy")["depends"]
[..., 'numpy-base 1.18.5.*', ...]
# Which depends on MKL and openblas (amongst other things).
>>> search_distribution("numpy-base")["depends"]
['blas 1.0 mkl', 'icc_rt >=2019.0.0', 'mkl >=2019.4,<2021.0a0', 'mkl-service >=2,<3.0a0', 'python >=3.7,<3.8.0a0', 'vc >=14.1,<15.0a0', 'vs2015_runtime >=14.16.27012,<15.0a0']
# Which contain the DLLs we're after.
>>> search_distribution("mkl")["files"]
['Library/bin/libimalloc.dll', 'Library/bin/mkl_avx.dll', 'Library/bin/mkl_avx2.dll', 'Library/bin/mkl_avx512.dll', 'Library/bin/mkl_blacs_ilp64.dll', '...]
```

So, without processing a package's requirements, we get nothing.

File: PyInstaller/utils/hooks/conda.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+module should therefore be guarded by a try/except clause::
+
+    try:
+        from PyInstaller.utils.hooks import conda_support
+    except ImportError:
+        # Not using conda.
+        pass
+    else:
+        # Code goes here. e.g.
+        binaries = conda_support.collect_dynamic_libs("numpy")
```
[the remark this answers is not in the document]
Comment (bwoodsend) [reply PyInstaller/utils/hooks/conda.py:31]: They're not quite equivalent. If a conda user decides they'd rather use `venv` (or something similar) then this would break in that environment because `is_conda` would be true, but the environment wouldn't have the `conda-metda` dir. But I was going to use @rokm's suggestion of adding a `compat.is_pure_conda` which would allow for a plain `if` statement instead of a `try: expect`.

**Comment by bwoodsend** [comment]
I promise I will get back to this at some point. Please hold off reviewing for now...

File: PyInstaller/utils/hooks/conda.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    """
+    # Locate the right json. The name format is:
+    #   [distribution name]-[version][ugly build # hash].json
+
+    # The json name can't be determined exactly so they must be glob searched:
+    for path in CONDA_META_DIR.glob(name + "-*.json"):
+        # But this will include distributions with a prefix of 'name'.
+        # e.g. 'numpy' would also pick up 'numpy-base'.
+        dic = json.loads(path.read_text())
+        if name == dic["name"]:
```
Comment (Legorooj) [inline PyInstaller/utils/hooks/conda.py:82]: Just a slight problem...
```suggestion
        if name == dic[name]:
```

File: PyInstaller/utils/hooks/conda.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            # Read the json to check we've found the right one.
+            # XXX: Maybe we should make this case and '-'/'_' insensitive?
```
Comment (Legorooj) [inline PyInstaller/utils/hooks/conda.py:84]:
```suggestion
            # TODO: Maybe we should make this case and '-'/'_' insensitive?
```

File: PyInstaller/utils/hooks/__init__.py [hunk 7]
```diff
@@ -1050,6 +1050,18 @@ def collect_all(package_name, include_py_files=True):
...
+if is_conda:
+    from . import conda as conda_support
+    if not conda_support.CONDA_META_DIR.exists():
+        logger.warning(
+            "Assuming this isn't an Anaconda environment or an additional "
+            "venv/pipenv/... environment manager is being used on top "
+            "because the conda-meta folder %s doesn't exist."
+            % conda_support.CONDA_META_DIR
+        )
+        del conda_support
```
[the remark this answers is not in the document]
Comment (Legorooj) [reply PyInstaller/utils/hooks/__init__.py:1062]: Agreed. No need to delete it IMO.

**Review by Legorooj — CHANGES_REQUESTED** [review]

**Comment by Legorooj** [comment]
After that and the CI is fixed, this is ready to go.

File: PyInstaller/utils/hooks/conda.py (hunk 5 above)
Comment (bwoodsend) [reply PyInstaller/utils/hooks/conda.py:82]: No, actually we do want to check if the `"name"` attribute in the json matches the `name` we are looking for. So the quotes are deliberate. I'll chuck in a comment.

**Comment by bwoodsend** [comment]
So far I've transformed what I had written into something which roughly mirrors `importlib.metadata`. I'm going to move it to the `building` folder and have a go at slotting its usage into the `analysis.py`.

File: PyInstaller/utils/hooks/conda.py [hunk 8]
```diff
@@ -121,8 +121,7 @@ def walk_dependency_tree(initial, excludes=None):
             logger.warning(
                 "Conda distribution '%s', dependency of '%s', was not found. "
                 "If you installed this distribution with pip then you may "
-                "ignore this warning." % (name, initial)
-            )
+                "ignore this warning." % (name, initial))
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:232]: Please abstain from non-functional changes like this one. See development guideline for rational.

File: PyInstaller/utils/hooks/conda.py [hunk 9]
```diff
@@ -155,8 +154,10 @@ def walk_dependency_tree(initial, excludes=None):
     lib_dir = Path("lib")
 
 
-def collect_dynamic_libs(name, dest=".",
-                         recurse_dependencies=True, excludes=None):
+def collect_dynamic_libs(name,
+                         dest=".",
+                         recurse_dependencies=True,
+                         excludes=None):
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py]: non-functional changes

File: PyInstaller/utils/hooks/conda.py [hunk 10]
```diff
@@ -39,11 +39,11 @@
 import sys
 from pathlib import Path
 import json
+import fnmatch
 
 from PyInstaller import compat
 from PyInstaller.log import logger
 
-
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py]: non-functional change

File: PyInstaller/utils/hooks/conda.py [hunk 11]
```diff
@@ -193,3 +194,67 @@ def collect_dynamic_libs(name, dest=".",
...
+
+    :param file: A Python filename relative to Conda root (sys.prefix).
+    :return: Package name or None.
+
+    This function only considers single file packages e.g. ``foo.py`` or
+    top level ``foo/__init__.py``\\ s. Anything else is ignored (returning
+    ``None``).
+    """
+    file = Path(file)
+    if file.name == "__init__.py":
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:366]: This fails if the `__init__` module is an extension module. See 0b496be42c67591576c1943635083ee6bc2940da
```suggestion
    if file.stem == "__init__" and file.suffix in compat.ALL_SUFFIXES:
```

File: PyInstaller/utils/hooks/conda.py [hunk 12]
```diff
@@ -193,3 +194,67 @@ def collect_dynamic_libs(name, dest=".",
...
+        file = file.parent
+
+    elif not (file.suffix in compat.ALL_SUFFIXES or file.is_dir()):
+        return
+
+    for prefix in PYTHONPATH_PREFIXES:
+        if len(file.parts) != len(prefix.parts) + 1:
+            continue
+        if fnmatch.fnmatch(str(file.parent), str(prefix)):
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:385]: Why using fnmatch here? Please add a short comment explaining.
Doesn't the following work?
```
if file.parent.parts == prefix.parts:
```

File: PyInstaller/utils/hooks/conda.py [hunk 13]
```diff
@@ -193,3 +194,67 @@ def collect_dynamic_libs(name, dest=".",
...
+            return file.parts[len(prefix.parts)].split(".")[0]
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:386]: I assume `".`" should be `os.extsep`? Use `Path.stem` here?

File: PyInstaller/utils/hooks/conda.py [hunk 14]
```diff
@@ -105,7 +213,7 @@ def walk_dependency_tree(initial, excludes=None):
 
     # Rather than use true recursion, mimic it with a to-do queue.
     from collections import deque
-    distributions_done = {}
+    done = {}
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:216]: Also here: please abstain from non-functional changes.

File: PyInstaller/utils/hooks/conda.py [hunk 15]
```diff
@@ -145,29 +250,68 @@ def walk_dependency_tree(initial, excludes=None):
...
 else:
-    lib_dir = Path("lib")
+    lib_dir = PackagePath("lib")
 
 
-def collect_dynamic_libs(name,
-                         dest=".",
-                         recurse_dependencies=True,
-                         excludes=None):
+def collect_dynamic_libs(name, dest=".", dependencies=True, excludes=None):
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:304]: Please abstain from non-functional changes. Isn't this just reverting part of the last commit?!

File: PyInstaller/utils/hooks/conda.py [hunk 16]
```diff
@@ -181,80 +325,101 @@ def collect_dynamic_libs(name,
...
     return files
 
 
+# --- Map packages to distributions and vice-versa ---
+
 PYTHONPATH_PREFIXES = []
-for path in sys.path:
-    path = Path(path)
+for _path in sys.path:
+    _path = Path(_path)
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:341]: non-functional change

File: PyInstaller/utils/hooks/conda.py [hunk 17]
```diff
@@ -181,80 +325,101 @@ def collect_dynamic_libs(name,
...
     try:
-        PYTHONPATH_PREFIXES.append(path.relative_to(sys.prefix))
+        PYTHONPATH_PREFIXES.append(_path.relative_to(sys.prefix))
     except ValueError:
         pass
 
+PYTHONPATH_PREFIXES.sort(key=lambda p: len(p.parts), reverse=True)
+
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:348]: This is a run-time constant. Please more near the top.

File: PyInstaller/utils/hooks/conda.py [hunk 18]
```diff
@@ -181,80 +325,101 @@ def collect_dynamic_libs(name,
...
     :return: Package name or None.
 
     This function only considers single file packages e.g. ``foo.py`` or
     top level ``foo/__init__.py``\\ s. Anything else is ignored (returning
     ``None``).
     """
-    file = Path(file)
+    # TODO: In theory, a package that doesn't have an ``__init__.py`` would be
+    #       missed. I've never seen a published package which didn't have an
+    #       ``__init__.py`` but it's not mandatory so it could happen.
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:363]: When I first read the comment, I was confused. The only reason a package not having an `__init__` module are PEP 420 namespace packages. In the comment, please refer to these to avoid confusion, something like this:
```
TODO: Handle PEP 420 namespace packages (which are missing `__init__` module). Anyhow
no Conda PEP 420 namespace packages are known.
```

File: PyInstaller/utils/hooks/__init__.py [hunk 19]
```diff
@@ -1050,6 +1050,17 @@ def collect_all(package_name, include_py_files=True):
...
 
+if is_pure_conda:
+    from . import conda as conda_support  # noqa: F401 (test import)
+elif is_conda:
+    from .conda import CONDA_META_DIR as _tmp
+    logger.warning(
+        "Assuming this isn't an Anaconda environment or an additional venv/"
+        "pipenv/... environment manager is being used on top because the "
+        "conda-meta folder %s doesn't exist." % _tmp
+    )
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/__init__.py:1061]:
```suggestion
    )
    del _tmp
```

File: PyInstaller/utils/hooks/conda.py [hunk 20]
```diff
@@ -181,80 +325,101 @@ def collect_dynamic_libs(name,
...
-                distributions_by_package_name[pkg] = distribution
+        dist = Distribution(path)
+        distributions[dist.name] = dist
 
-    return distributions_by_package_name
+    return distributions
 
 
-distributions_by_package_name = _make_package_to_distribution_map()
+distributions = _init_distributions()
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:412]: Please use the new name in the prio commit already, to reduce diffs. Same rational as for non-functional changes.

File: PyInstaller/utils/hooks/conda.py [hunk 21]
```diff
@@ -44,59 +46,165 @@
 from PyInstaller import compat
 from PyInstaller.log import logger
 
+if compat.is_py38:
+    from importlib.metadata import PackagePath as _PackagePath
+else:
+    from importlib_metadata import PackagePath as _PackagePath
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/conda.py:52]: This needs to be added to setup.cfg and requirements.txt, too.

**Review by htgoebel — CHANGES_REQUESTED** [review]
Here are some inline comments. Sorry for being pick on non-functional changes. See #2727 for rational.

**Comment by htgoebel** [comment]
Thanks for your answers, this clarified things.
**Beside of the nit-picks (so inline comments), this PR is okay from my POV.**

> > If conda changes the filename, how can the extension module be imported at all in Conda?
> 
> I'm not sure what you're referring to. Extension modules aren't renamed. DLLs are moved and some new DLLs become dependencies - that's all.

IC. So let me rephrase:
If conda changes the filename, how is the *DLL* found in Conda? Are they changing the DLL-name in the dependent, too?

> > how we could get rid of adding a hook for every package
> 
> […] Only those that Conda have actively altered. […] Only the big libraries get the special Conda treatment. That all being said - if we can make this automatic, I definitely think we should.

Okay, so we hopefully can live with the need of creating hooks for these packages..

> > whether this could be written to be more general and support non-conda installations, too
> 
> Do you mean, _could we do something similar using `importlib.metdata` to find DLL dependencies when not using Conda_?

No exactly: I was wondering whether we actually could use existing `importlib.metdata` instead of re-implementing it. I spotted your change using this package, so I'm confident you already took care about this.

File: PyInstaller/utils/hooks/conda.py (hunk 10 above)
Comment (Legorooj) [reply PyInstaller/utils/hooks/conda.py]: IMO removing a blank line isn't actually an issue - it doesn't show up in the blame.

Comment (Legorooj) [reply PyInstaller/utils/hooks/conda.py]: And anyway, You're reviewing a single commit - it's better to wait until the PR is done, and then nitpick on non-functional changes after all the functional ones are made. (And often commits get squashed together, eliminating some changes totally)

File: PyInstaller/utils/hooks/__init__.py (hunk 19 above)
Comment (Legorooj) [reply PyInstaller/utils/hooks/__init__.py:1061]: This isn't needed - `gc` will clean it up automatically (due to scopes).

**Comment by bwoodsend** [comment]
I'm mostly posting this for my own reference. The data files kept in Conda's `share` folder aren't all junk as I previously assumed (although it's definitely mostly junk or we'd need to include it more often). See [here](https://github.com/pyinstaller/pyinstaller-hooks-contrib/blob/5cad909a07ddc56510868672f008185ebdb2577d/src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-pyproj.py#L44-L49) where they're actually needed. This would imply that we will also need a `collect_data_files()` Conda equivalent.

**Comment by htgoebel** [comment]
@bwoodsend I'd like to have this pull-request in 4.2, which I plan to release on 2020-12-31. Are you going to finish this pull-request?

File: PyInstaller/utils/hooks/__init__.py (hunk 19 above)
Comment (htgoebel) [reply PyInstaller/utils/hooks/__init__.py:1061]: Unfortunately this will not be garbage-collected: The module globals will be kept as long as any module has a reference to this module (g.g due to some import).

**Comment by bwoodsend** [comment]
That's the nit-picks dealt with. There's a new commit f0c430562d19cf4cb909f8fdb531733a61e8d736 which runs the `conda.collect_dynamic_libs()` on every package being included so we should not need to write any hooks for Conda packages.

File: PyInstaller/building/build_main.py [hunk 22]
```diff
@@ -570,6 +572,46 @@ def _check_python_library(self, binaries):
...
+            # (dest_name, source, 'DATA') format.
+            for (source, dest) in dlls:
+                # I'm intentionally skipping the DLL dependency scanner (by
+                # adding binary files as datas) for two reasons:
+                #   - Conda packages should have brought everything they use
+                #     with them so anything else detected would be system
+                #     libraries.
+                #   - There are litter files (.pdb, .def, ...) which, if
+                #     scanned, cause errors or uncatchable pop-up windows.
+                self.datas.append((os.path.basename(source), source, "DATA"))
```
Comment (bwoodsend) [inline PyInstaller/building/build_main.py:613]: This puts all DLLs in the package root. In the interest of decluttering the root and possibly to make clearer where DLLs come from, would anyone be opposed to putting them in a dedicated `conda-bin` folder and adding a runtime hook to add said `conda-bin` to `PATH`?

Comment (rokm) [reply PyInstaller/building/build_main.py:613]: Not sure how common conda is on macOS, but due to the library loader path rewriting magic we do on binaries there, this will mean you'll need to keep track of which dynlibs go where and adjust the path in the binaries that are linked against them. How much of a problem would that be?

**Comment by rokm** [comment]
Hmm, I've tested this with a simple `matplotlib` example, which ended up pulling in `PyQt5` (even though it did not collect `PyQt5` backend, which is an issue for another day...), and I now see a lot of utility executables being collected (because they were located in `Library\bin`, same location as the DLLs): `assistant.exe` (Qt Assistant app), `designer.exe` (Qt designer app), `sqlite3.exe`, `tclsh.exe`, `xz.exe`, etc. 

Collecting more than needed tends to be safer than collecting less, but those extra applications will likely cause complaints among users. 

Along the similar line of thought, we are now brute-force collecting all Qt5 DLLs, even though our PySide2/PyQt5 hooks go out of their way to ensure that only the required parts of Qt5 are collected.

**Comment by bwoodsend** [comment]
> I've tested this with a simple matplotlib example, which ended up pulling in PyQt5 (even though it did not collect PyQt5 backend

I'm not seeing this. For me, PyQt5 gets pulled in as a backend by default but goes away again if you `--exclude=PyQt5` (which is what normally happens). 

> I now see a lot of utility executables being collected (because they were located in Library\bin, same location as the DLLs): `assistant.exe`

Yeah, it's not pretty. Perhaps a simple `*.exe` filter would do the job but I'm reluctant to do this because that would knock out any embedded command line tools. Also, whilst it's a very visible form of junk, most of them are just shims so the increase in size is still less than the size of the pdb junk files and is peanuts compared to the 400MB of unused mkl dlls Conda ships. If they're not needed then they shouldn't be in Conda packages to begin with.

> Along the similar line of thought, we are now brute-force collecting all Qt5 DLLs, even though our PySide2/PyQt5 hooks go out of their way to ensure that only the required parts of Qt5 are collected.

If it does, then that means the DLLs are in the special `Library/bin` folder which is outside or PyInstaller's search so unless our PyQt5 hooks have special Conda support (which I don't think they do) then Conda PyQt5 + PyInstaller would fail anyway. Certainly the `get_qt_binaries()` doesn't work from my Conda environment. But I get what you're saying. Perhaps we could later provide a blacklist of packages which do their own Conda DLL collection. TBH, I wish Qt wasn't one huge package - I hate huge packages which then need to be picked apart.

**Comment by rokm** [comment]
> > I've tested this with a simple matplotlib example, which ended up pulling in PyQt5 (even though it did not collect PyQt5 backend
> 
> I'm not seeing this. For me, PyQt5 gets pulled in as a backend by default but goes away again if you `--exclude=PyQt5` (which is what normally happens).

Ah, right, rookie mistake on my part - I saw `_tkagg` and `_backend_agg` extension module files on filesystem and concluded the other were missing. But they are, of course, regular python modules and are as such collected into PYZ...

> > I now see a lot of utility executables being collected (because they were located in Library\bin, same location as the DLLs): `assistant.exe`
> 
> Yeah, it's not pretty. Perhaps a simple `*.exe` filter would do the job but I'm reluctant to do this because that would knock out any embedded command line tools. Also, whilst it's a very visible form of junk, most of them are just shims so the increase in size is still less than the size of the pdb junk files and is peanuts compared to the 400MB of unused mkl dlls Conda ships. If they're not needed then they shouldn't be in Conda packages to begin with.

The problem is that they are unnecessary from the perspective of a frozen python app, but not necessarily Conda itself. I think Conda aims to be a general-purpose package manager/vendor. So it makes sense for them to package for example Qt Designer executable in their `qt` package, in contrast to for example PyQt5 pip wheels which bundle only stuff that is actually needed by the python bindings. (While on the other hand, PySide2 pip wheel does bundle the utility programs as well).

> > Along the similar line of thought, we are now brute-force collecting all Qt5 DLLs, even though our PySide2/PyQt5 hooks go out of their way to ensure that only the required parts of Qt5 are collected.
> 
> If it does, then that means the DLLs are in the special `Library/bin` folder which is outside or PyInstaller's search so unless our PyQt5 hooks have special Conda support (which I don't think they do) then Conda PyQt5 + PyInstaller would fail anyway. Certainly the `get_qt_binaries()` doesn't work from my Conda environment. But I get what you're saying. Perhaps we could later provide a blacklist of packages which do their own Conda DLL collection. TBH, I wish Qt wasn't one huge package - I hate huge packages which then need to be picked apart.

The DLLs are in the `Library\bin` folder, but the necessary subset is still picked up correctly by PyInstaller, thanks to the dependency analysis:

```
...
7375 DEBUG: Analyzing c:\users\rok\anaconda3\envs\pyi-pyqt5\lib\site-packages\PyQt5\QtCore.pyd
7389 DEBUG: Adding Qt5Core.dll dependency of QtCore.pyd from C:\Users\Rok\anaconda3\envs\pyi-pyqt5\Library\bin\Qt5Core.dll
...
```

So perhaps the issue is not that `Library\bin` is not in the search path, but rather that some Conda packages (e.g., `numpy`) have DLL dependencies that are loaded at runtime (as opposed to being directly linked) and therefore slip through our dependency analysis (e.g., `libiomp5md.dll`). If this is the case, it would make more sense to perform this conda-metadata-based binary collection only for those problematic packages, instead of all of them.

**Comment by bwoodsend** [comment]
> The DLLs are in the Library\bin folder, but the necessary subset is still picked up correctly by PyInstaller, thanks to the dependency analysis:

Hmm, you're right. Our PyQt hooks are Conda proof (albeit only just). What do you reckon then? Give hooks (like Qt) an option to opt out of having their DLLs dumped in or just drop the last commit entirely and have hooks opt in using the original `if is_pure_conda: binaries = conda.collect_dynanic_libs("numpy")`? I'd say I'm leaning towards the second.

**Comment by rokm** [comment]
> Hmm, you're right. Our PyQt hooks are Conda proof (albeit only just). What do you reckon then? Give hooks (like Qt) an option to opt out of having their DLLs dumped in or just drop the last commit entirely and have hooks opt in using the original `if is_pure_conda: binaries = conda.collect_dynanic_libs("numpy")`? I'd say I'm leaning towards the second.

I'm leaning towards the original opt-in idea as well.

Perhaps we could start with that, opt in the `numpy` package (which seems to be the source of majority of our Conda-induced issues), and then reassess the situation?

**Comment by bwoodsend** [comment]
OK, final commit is gone. I've stashed in a separate [add-all-conda-dlls](https://github.com/bwoodsend/pyinstaller/tree/add-all-conda-dlls) branch - perhaps we can use it as a quick test if we suspect an issue is Conda DLL based but aren't sure.

**Review by rokm — APPROVED** [review]
Looks good.

I tried to opt in the `numpy` package, and while there's still some unnecessary stuff collected from `Library\bin` (e.g., `.pdb`, `.exe`, `.pl` files), we can now deal with those on per-package basis in the hook by filtering the output of `conda.collect_dynamic_libs`).

**Review by Legorooj — APPROVED** [review]
lgtm!

File: PyInstaller/building/build_main.py (hunk 22 above)
Comment (htgoebel) [reply PyInstaller/building/build_main.py]: I'd prefer to put all extensions and dlls into subfolder, see #2468. So for this pull-request, I'd not touch this, rather finish this on :-)

File: PyInstaller/utils/hooks/__init__.py [hunk 23]
```diff
@@ -1131,6 +1131,18 @@ def collect_all(
...
 
 
+if is_pure_conda:
+    from . import conda as conda_support  # noqa: F401
+elif is_conda:
+    from .conda import CONDA_META_DIR as _tmp
+    logger.warning(
+        "Assuming this isn't an Anaconda environment or an additional venv/"
+        "pipenv/... environment manager is being used on top because the "
+        "conda-meta folder %s doesn't exist." % _tmp
```
Comment (htgoebel) [inline PyInstaller/utils/hooks/__init__.py]: Don't format strings for logging - this is done by the logger and only if the record as actually dumped. In our case this might be not relevant, since we run with "warning" level at least, but for other application this might matter.

**Review by htgoebel — APPROVED** [review]
I removed some white-space too much. Otherwise lgtm.

Please also not my comment about how to use logging.

File: PyInstaller/utils/hooks/__init__.py (hunk 23 above)
Comment (bwoodsend) [reply PyInstaller/utils/hooks/__init__.py]: Oops, sorry. You've told me that one before :-).
